### PR TITLE
Fixed #29850 -- Increased window range support.

### PR DESCRIPTION
**Pull request by sarahboyce**
Ticket: https://code.djangoproject.com/ticket/29850

- [X] Support for positive integer `start` and negative integer `end` for `RowRange`.
- [X] Support for `EXCLUDE CURRENT ROW`, `EXCLUDE GROUP`, `EXCLUDE TIES`, `EXCLUDE NO OTHERS`

(I will push up tweaks while I also figure it out better in my head)

### CODE REVIEW

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/base/operations.py [hunk 1]
```diff
@@ -714,42 +714,68 @@ def subtract_temporals(self, internal_type, lhs, rhs):
...
             raise NotSupportedError("This backend does not support window expressions.")
-        return self.window_frame_start(start), self.window_frame_end(end)
+        elif isinstance(start, int) and isinstance(end, int) and start > end:
+            raise ValueError("start cannot be greater than end.")
+        return self.window_frame_rows_start(start), self.window_frame_rows_end(end)
 
     def window_frame_range_start_end(self, start=None, end=None):
-        start_, end_ = self.window_frame_rows_start_end(start, end)
+        if not self.connection.features.supports_over_clause:
+            raise NotSupportedError("This backend does not support window expressions.")
```
Comment (felixxm) [inline django/db/backends/base/operations.py:775]: Do you think it's worth adding? `WindowFrame` subclasses are the only one that call this method and they can only be used with `Window` which already raises an exception when `supports_over_clause` is `False`.

File: django/db/backends/base/operations.py [hunk 2]
```diff
@@ -714,42 +714,68 @@ def subtract_temporals(self, internal_type, lhs, rhs):
...
+        return as_sql
+
+    def window_frame_range_start(self, start):
+        as_sql = self.window_frame_value(start, none_value=self.UNBOUNDED_PRECEDING)
+        if isinstance(start, int) and start > 0 or as_sql is None:
+            raise ValueError(
+                "start argument must be a negative integer, zero, or None, "
+                "but got '%s'." % start
+            )
+        return as_sql
```
Comment (felixxm) [inline django/db/backends/base/operations.py:752]: I'm not sure if these hooks are needed, I'd move the logic to the `window_frame_range_start_end()` (the same with `window_frame_rows_end()` and `window_frame_rows_start()`). As far as I'm aware, we don't need such a small granularity.

File: django/db/backends/base/operations.py (hunk 1 above)
Comment (sarahboyce) [reply django/db/backends/base/operations.py]: Ah clever - you're right, it's not required 👍

File: django/db/backends/base/operations.py (hunk 2 above)
Comment (sarahboyce) [reply django/db/backends/base/operations.py]: Thank you, it looks much nicer now

File: django/db/backends/base/operations.py [hunk 3]
```diff
@@ -714,42 +714,51 @@ def subtract_temporals(self, internal_type, lhs, rhs):
...
-        elif end is None:
-            return self.UNBOUNDED_FOLLOWING
-        raise ValueError(
-            "end argument must be a positive integer, zero, or None, but got '%s'."
-            % end
-        )
+            else:
+                return "%d %s" % (value, self.FOLLOWING)
+        elif value is None:
+            return unbound
```
Comment (felixxm) [inline django/db/backends/base/operations.py:726]: We raise `ValueError` before calling this method so we can reach this code only with `None`
```suggestion
        return unbound
```

Comment (felixxm) [reply django/db/backends/base/operations.py:726]: I wonder if it's worth passing `unbound`, maybe:
```python
start_ = self.window_frame_value(start) or self.UNBOUNDED_PRECEDING
end_ = self.window_frame_value(end) or self.UNBOUNDED_FOLLOWING
```

File: docs/ref/models/expressions.txt [hunk 4]
```diff
@@ -954,9 +976,9 @@ with the average rating of a movie's two prior and two following peers:
 
 If the database supports it, you can specify the start and end points based on
 values of an expression in the partition. If the ``released`` field of the
-``Movie`` model stores the release month of each movies, this ``ValueRange``
+``Movie`` model stores the release month of each movie, this ``ValueRange``
 example annotates each movie with the average rating of a movie's peers
-released between twelve months before and twelve months after the each movie:
+released between twelve months before and twelve months after each movie:
```
Comment (felixxm) [inline docs/ref/models/expressions.txt:981]: These typos should be fixed separately.

File: docs/ref/models/expressions.txt [hunk 5]
```diff
@@ -970,6 +992,13 @@ released between twelve months before and twelve months after the each movie:
     ...     ),
     ... )
 
+
```
Comment (felixxm) [inline docs/ref/models/expressions.txt:995]:
```suggestion
```

File: docs/releases/5.1.txt [hunk 6]
```diff
@@ -171,6 +171,14 @@ Models
 * :meth:`.QuerySet.explain` now supports the ``generic_plan`` option on
   PostgreSQL 16+.
 
+* :class:`~django.db.models.expressions.RowRange` now accept positive integers
+  for the ``start`` argument and negative integers for the ``end`` argument.
+
+* The new ``exclusion`` argument of
+  :class:`~django.db.models.expressions.RowRange` and
+  :class:`~django.db.models.expressions.ValueRange` will add an exclusion
```
Comment (felixxm) [inline docs/releases/5.1.txt:179]:
```suggestion
  :class:`~django.db.models.expressions.ValueRange` add an exclusion
```

File: docs/ref/models/expressions.txt [hunk 7]
```diff
@@ -911,6 +911,25 @@ Both classes return SQL with the template:
 
     %(frame_type)s BETWEEN %(start)s AND %(end)s
 
+
+.. class:: Exclusion
```
Comment (felixxm) [inline docs/ref/models/expressions.txt:915]: `versionadded` annotation is missing.

File: docs/releases/5.1.txt [hunk 8]
```diff
@@ -171,6 +171,14 @@ Models
 * :meth:`.QuerySet.explain` now supports the ``generic_plan`` option on
   PostgreSQL 16+.
 
+* :class:`~django.db.models.expressions.RowRange` now accept positive integers
```
Comment (felixxm) [inline docs/releases/5.1.txt:174]:
```suggestion
* :class:`~django.db.models.expressions.RowRange` now accepts positive integers
```

**Comment by felixxm** [comment]
@sarahboyce Thanks :+1: I moved the first commit to the separate PR, #17422.

**Comment by sarahboyce** [comment]
Superseded by https://github.com/django/django/pull/17423

**Pull request closed without merging**